File: persona1_java/src/cc4p1/raft/RaftNodeMain.java
package cc4p1.raft;

import java.util.concurrent.TimeUnit;

/**
 * Punto de entrada del Nodo Raft #1 (Java).
 *
 * Uso:
 *   java cc4p1.raft.RaftNodeMain <selfId> <rutaClusterProperties>
 *
 * Ejemplo:
 *   java cc4p1.raft.RaftNodeMain java1 config/cluster.properties
 */
public class RaftNodeMain {

    public static void main(String[] args) throws Exception {
        if (args.length < 2) {
            System.err.println("Uso: RaftNodeMain <selfId> <rutaClusterProperties>");
            System.exit(1);
        }
        String selfId = args[0];
        String configPath = args[1];

        RaftConfig config = new RaftConfig(configPath, selfId);
        StateMachine stateMachine = new StateMachine();
        RaftNode node = new RaftNode(config, stateMachine);
        RaftServer server = new RaftServer(config.selfPort, node);

        Thread serverThread = new Thread(server, "raft-server-" + selfId);
        serverThread.setDaemon(false);
        serverThread.start();

        node.start();

        // Hilo separado que imprime el estado del nodo cada 5s (util para
        // demostrar en vivo la eleccion de lider y la replicacion del log).
        Thread statusThread = new Thread(() -> {
            while (true) {
                try {
                    TimeUnit.SECONDS.sleep(5);
                    System.out.println(node.describeStatus());
                } catch (InterruptedException e) {
                    return;
                }
            }
        }, "status-" + selfId);
        statusThread.setDaemon(true);
        statusThread.start();

        Runtime.getRuntime().addShutdownHook(new Thread(server::stop));
    }
}

File: persona2_python/src/raft/raft_node_main.py


import sys
import threading
import time

from raft.raft_config import RaftConfig
from raft.raft_node import RaftNode
from raft.raft_server import RaftServer
from raft.state_machine import StateMachine


def _status_loop(node: RaftNode):
    while True:
        time.sleep(5)
        print(node.describe_status())


def main():
    if len(sys.argv) < 3:
        print("Uso: python raft_node_main.py <selfId> <rutaClusterProperties>")
        sys.exit(1)

    self_id = sys.argv[1]
    config_path = sys.argv[2]

    config = RaftConfig(config_path, self_id)
    state_machine = StateMachine()
    node = RaftNode(config, state_machine)
    server = RaftServer(config.self_port, node)

    server_thread = threading.Thread(target=server.serve_forever, name="raft-server-%s" % self_id, daemon=False)
    server_thread.start()

    node.start()

    status_thread = threading.Thread(target=_status_loop, args=(node,), daemon=True)
    status_thread.start()

    try:
        server_thread.join()
    except KeyboardInterrupt:
        server.stop()
        node.stop()


if __name__ == "__main__":
    main()
